### ISSUE TITLE
Input Data Resources

### ISSUE THREAD

**Issue by jesusjuansalgado** [ivoa-std/ExecutionPlanner#39]
As part of the request, it is crucial for projects like SKAO to have a list of input data resources used by the workflow. This is a very relevant factor to decide the correct node to be used for the execution.
One options would be to add a new subsection at 12.2.X like
Data Location
The execution request could define the input data to be used for the workflow execution so the server can evaluate if the data is accessible or colocated. This could be done by providing metadata entries with these metadata:

obs_collection | DataID.collection 
obs_id | DataID.observationID
access_url | Access.reference

in line with IVOA ObsCore.
Using this metadata, a summary could be provided with an accessibility rank (TBD)

What do you think? I could take care of drafting this part.

**Comment by Zarquan** [comment]
Yep, I agree. I had conflated data with storage. My assumption was that a data object would be referenced as a storage resource, which is probably wrong.

Following your suggestion, we could have "Data resources" at the same level in the data model as "Compute resources" (section 12.2.7), between "Storage resources" (section 12.2.9) and "Authentication" (section 12.2.10).

```
resources:
  compute:
    ....
  storage:
    ....
  data:
    - name: input-dataset
      type: http://..../data-types/ivoa-dataset
      spec:
        # IVOA metadata
```

Using the type URI `http://...data-types/ivoa-data` means we can use IVOA specific metadata in this section, while still allowing other projects outside the IVOA to use their own types of metadata.

```
resources:
    ....
  data:
    - name: tree-cover-dataset
      type: http://..../data-types/forestry-dataset
      spec:
        # Forestry specific metadata
```

**Comment by Zarquan** [comment]
Importing comments from email thread, on 2023-10-24 12:09, Bertocco, Sara wrote:
> Hi,
> for what concerns data, I would like to share my experience, probably Dave
> already has the same knowhow but I do not know about the others. In the
> grid middleware gLite, the Job Description Language was used to describe
> the job(= "executable thing"). It was containing the definition of an
> InputSandbox and OutputSandbox parameters. Simple descriptions can be found:
> https://indico.cern.ch/event/422979/contributions/1036845/attachments/892335/1256365/JDL.pdf
> It is the same job description language used by DIRAC, see for example
> https://dirac.readthedocs.io/en/latest/UserGuide/Tutorials/JDLsAndJobManagementBasic/index.html
> And I cite myself also, see the attachment, searching "jdl" you find a more
> complex and explicative example.
> I think we need something similar and probably this is the reason why
> Raymond Oonk cited Dirac.
> Cheers, Sara

**Comment by Zarquan** [comment]
Importing comments from email thread, on 2023-10-24 12:30, Salgado, Jesus wrote:
> Hi Sara,
> Thanks for this! I think the solution I am proposing is more adequate
> for the IVOA.
> You can just refer that your workflow is making use of a particular
> data collection, e.g. CHANDRA/ACIS-S
> or going to the exact definition of all the data inputs (that could be
> nice but a killer for users but the normal way that most of the
> planners, like DIRAC, make use)
> 
> There is a note by Markus and Mark
> https://www.ivoa.net/documents/discovercollections/20190520/EN-discovercollections-1.1-20190520.html
> 
> but I think the problem is that you need registration in this approach.
> In the case of the execution planner, as it is a realtime decision
> from the node, I think this will be more adequate and powerful.
> Basically, if the user knows she/he is going to use, e.g. Euclid data
> (not knowing the exact data products access URLs because I have some
> Obscore queries integrated in my workflow) the user can make use the
> planner (setting the obs_collection entry into the “inputsandbox”) to
> identify which nodes are able to handle Euclid data (and with certain
> score as the data could be collocated or only accesible).
> 
> Of course, the exact definition of the data products is better but the
> different granularity levels of definition could be more usable in our
> use case. Just define the “inputsandbox” as records with all (or some)
> of the metadata:
> 
> obs_collection | DataID.collection
> obs_id | DataID.observationID
> access_url | Access.reference
> 
> defined.
> 
> Do you know if InputSandBox is compulsory for Dirac?
> 
> Cheers,
> Jesus

**Comment by Zarquan** [comment]
Importing comments from email thread, on 2023-10-24 13:19, Salgado, Jesus wrote:
> Also, could it be that the exact input data products are more related
> to the execution worker that to the execution planner where a more
> ambiguous definition could be enough?

**Comment by Zarquan** [comment]
Importing comments from email thread, on 2023-10-24 16:00, Bertocco, Sara wrote:
> Yes, could be, but in my opinion, the ExecutionPlanner needs some
> information about data to be processed to know if the "underlined computing
> node" is able to access the required data (locally or through a defined
> protocol).
> Sara

**Comment by Zarquan** [comment]
Importing comments from email thread, on 2023-10-24 16:27, Salgado, Jesus wrote:
> True, this is what I tried to say in “certain score as the data could
> be collocated or only accessible”  (it could be even non-accessible).
> The score should be able to have a TBD level of granularity (ideally,
> we could even put the expected data latency but this could be too much
> for a general use… maybe for the SRCNet)

**Comment by Zarquan** [comment]
Yes, a section on data resources makes sense. Probably at the same level as compute and storage resources.

Caveats. We need to start with a generic non-astronomy definition:

```
resources:
  data:
  - name: !!str # A name for the data resource.
    type: !!str # A URI to identify the type of data resource.
    spec:
      # Type specific details about the data resource.
```

and then add type specific extensions for IVOA data resources using the `type:` and `spec:` pattern.

```
resources:
  data:
  - # Type URI for an IVOA data resource.
    type: http://.../data-resource-types/ivoa-data-resource
    spec:
      # Type specific details about an IVOA data resource,
```

However, before we can add much more detail here I think we need to think about what an IVOA data resource means.

> You can just refer that your workflow is making use of a particular
> data collection, e.g. CHANDRA/ACIS-S

> ... if the user knows she/he is going to use, e.g. Euclid data
> (not knowing the exact data products access URLs because I have some
> Obscore queries integrated in my workflow) the user can make use the
> planner (setting the obs_collection entry into the “inputsandbox”) to
> identify which nodes are able to handle Euclid data (and with certain
> score as the data could be collocated or only accesible).

This feels very much like data discovery, which I deliberately excluded from the ExecutionPlanner specification, see section 3.1.

Data discovery is not a solved problem, and my concern is that including it in the ExecutionPlanner specification it would delay the process of getting to version 1.0. It would also embed a lot of project specific things into the ExecutionPlanner specification.

I am fairly confident that I have solutions for most if not all of the incomplete sections in the current ExecutionPlanner draft. I just need time to actually write them down with clear examples and descriptions.

Data discovery, on the other hand, is a dark unknown that I would like someone else to solve.

I would prefer to have IVOA data discovery in a separate document, that describes what terms the user would search for, what results the data discovery service (registry?) would provide, and how those results could be used to determine proximity. Then, the `ivoa-data-resource' section of the ExecPlanner document could refer to the IVOA data discovery document for details.

```
resources:
  data:
  - name: my-ivoa-data-resource
    # Type URI for an IVOA data resource.
    type: http://.../data-resource-types/ivoa-data-resource
    spec:
      # IVOA specific data discovery parameters that
      # resolve to one or more IVOA data resource identifiers.
      # See [IVOA data discovery document] for details.
```

Bear in mind that the original request was from SKAO:
> it is crucial for projects like SKAO to have a list of input data resources used by the workflow

In which case it might be relevant to work on an SKA specific extension, using SKA data references.

As with IVOA data discovery, the details of SKA data discovery would be in a separate document that describes how the
SKA data discovery service works. What terms the user would search for, what results the discovery service would provide, and how those results could be used to determine proximity.

Then, the `ska-data-resource' section of the ExecutionPlanner document could refer to the SKA data discovery document
for details.

```
resources:
  data:
  - name: my-ska-data-resource
    # Type URI for a SKA data resource.
    type: http://.../data-resource-types/ska-data-resource
    spec:
      # SKA specific data discovery parameters that
      # resolve to one or more SKA data resource identifiers.
      # See [SKA data discovery document] for details.
```

**Comment by Zarquan** [comment]
TL;DR;

* I agree we need a section for data resources.
* I would prefer not to include details of data discovery in the ExecutionPlanner specification.
* I would like to avoid including IVOA or SKA specific details in the ExecutionPlanner specification.
* It would be better to create separate documents that describe data discovery in the context of the IVOA and SKA.
* We can then include `ivoa-data-resource` and `ska-data-resource` sections in the ExecutionPlanner data model that refer to the external documents for details.

**Comment by jesusjuansalgado** [comment]
The main reason of the interest of the SRCNet in promoting the ExecutionPlanner is to discover which nodes are able to execute a particular workflow because they have the data colocated. Without this, the spec is not really useful for us and for the other astronomical science platforms that are handling with big data sets and data transfer latencies. If we implement something like an ExecutionPlanner, it will have the input data definition for sure (this is why, it is better to standardise it). I understand that adding a new concept could complicate the standard but if not having it produces that it is not really usable, it is a main problem to me. So, in summary, the data location is not a reason to implement an ExecutionPlanner is THE reason to implement something like this.

Leaving apart the SRCNet case that I commented, I think the problem is shared by many.
One requirement we share with the main astronomical science analysis platforms is to do not misuse resources to analyse in one location data from other missions/telescopes. For example, nobody wants at LSST that someone is making use of their system to analyse Euclid data. 

Resources are limited and expensive so this is applicable to the main astronomical projects of the planet.  It is not, in my view, just an SKA requirement but a very strong requirement for all the missions that have decided to create science analysis platforms because it is the optimal (but expensive) way to analyse their data (due to complexity or size).

People involved are, basically, the target audience of the ExecutionPlanner spec and this is one of the reasons why I have proposed to promote it spec as GWS chair. The integration with the (optional) definition of the input data is a main justification for the ExecutionPlanner to me. If this is not covered, science platforms engineers would need to explore different solutions.

About trying to define the standard without astronomical concepts (to define it usable by other communities) looks to me too optimistic as other communities could take the concept but they would adapt it (or just use other packages). The main reason for astronomical platforms to implement it is because it would be an IVOA standard and this will not be so effective for other communities. I think we can prepare a skeleton and create an extension on data location within the same spec to make it useful. Having a separate spec (and waiting for the standardisation) is overcomplicate the standardisation process with linked standards that, we know, do not work well in the IVOA. Having this as an extension inside the doc will allow to have many implementations in a short term and an easy extension to other communities if needed.

The collections discovery note (if this is what you mean about IVOA data discovery) it is a protocol that implies the use of the registry, so is it static. Accessibility to a particular data set is a realtime assessment (that could include not only colocated data but, also, an estimation of accessibility using, e.g. network bandwidth) so the ExecutionPlanner is a better approach to integrate something like this from the execution point of view.  To duplicate a realtime standard to ask for data accessibility so it could be decided if a certain workflow could be execute and then ask to the execution planner about other characteristics seems to be overcomplicated when both things could be integrated.

**Comment by Zarquan** [comment]
**Section n IVOA extensions**
**Section n.1 ObsCore data resource**

A data resource that will parse a list of data objects in an IVOA ObsCore VOTable and apply the task to each data
object in the list.

```
resources:
  data:
  - type: https://..../data-resource-types/ivoa-obscore-resource
    spec:

      # Input parameters supplied by the client.
      # A URL pointing to the ObsCore VOTable.
      source-url: !!str
      # or, a base64 encoding of the ObsCore VOTable.
      source-data: !!str
      # Execution mode - a separate offer for each row, or one offer for the combined set.
      execution-mode: !!str [SEPARATE|COMBINED]

      # Response values in the offer supplied by the ExecutionPlanner.
      # An MD5SUM checksum of the VOTable file.
      source-md5: !!str
      # A unique identifier for the row if the execution mode is SEPARATE.
      row-ident: !!str
      # A calculation of the data proximity.
      data-proximity: !!float
      # A calculation of the execution cost.
      execution-cost: !!float
```

See https://ivoa.net/documents/Notes/IVOAExecutionPlannerExtensions for additional details about the ObsCore data resource works.

**Comment by Zarquan** [comment]
That is all that would be needed in the main ExecutionPlanner document. Everything else can be developed in the referenced IVOAExecutionPlannerExtensions note.

Under the hood, the ExecutionPlanner would use fields from the ObsCore data model to estimate the data proximity and execution cost for each object in the list, but how this is implemented should be covered in a separate document.

The reason for keeping it separate is that defining how to use the ObsCore fields to estimate data proximity and execution cost could easily run to several pages of very detailed IVOA specific explanation. Getting agreement on exactly what `data-proximity` and `execution-cost` mean may take a lot of work.

A pragmatic approach would be for the SKA team to go ahead and define a `ska-data-resource`, and the Rubin team can define a `rubin-data-resource`, including definitions of what terms like `data-proximity` and `execution-cost`
mean for their project. This would enable the projects to get up and running with prototypes now, and then we could work together to agree on an IVOA-ObsCore-data-resource later.

If we do end up with a simple, clear, and concise definition of `data-proximity` and `execution-cost` then by all means, we can incorporate it into the main ExecutionPlanner document.

**Comment by Zarquan** [comment]
If we do want to move the data discovery part into the ExecutionPlanner, then we could define a data resource for querying IVOA SimpleType services.

--------
**Section n IVOA extensions**
**Section n.2 SimpleService data resource**

````
resources:
  data:
  - type: https://..../data-resource-types/ivoa-simple-service
    spec:

      # Input parameters supplied by the client.
      # The service type
      service-type: !!str [SCS|SIA|SSA|SLAP]
      # The endpoint URL for the service.
      service-endpoint: !!str
      # or
      # The IVOA service registration for the service.
      service-ivorn: !!str
      # Execution mode - a separate offer for each row, or one offer for the combined set.
      execution-mode: !!str [SEPARATE|COMBINED]
      # A list of name-value params to pass to the service.
      service-params:
      - name:  "POS"
        value: "CIRCLE 12.0 34.0 0.5"
      - name:  "BAND"
        value: "500e-9 550e-9"

      # Response values in the offer supplied by the ExecutionPlanner.
      source-md5: !!str # An MD5SUM checksum of the ObsCore VOTable file.
      row-ident: !!str # A unique identifier for the row if the execution mode is SEPARATE.
      data-proximity: !!float # A calculation of the data proximity.
      execution-cost: !!float # A calculation of the execution cost.
````

The ExecutionPlanner uses the service params in the request to send a query to the IVOA SimpleService, and then applies the execution task to each data object in the SimpleService response.

See https://ivoa.net/documents/Notes/ExecutionPlannerExtensions for additional details about the SimpleService data resource works.

**Comment by Zarquan** [comment]
In the mean time, there is a lot we can do with what is already in the specification. One of the ideas I have been exploring is to indicate the data transfer costs using the `starttime` property of an offer.

Given the URL for a data object, the ExecutionPlanner can check to see if it already has a copy of the data object in a local cache. If not, and this is the first time it has this platform has downloaded this data object, it can use things like HTTP HEAD requests to estimate the data size and network speed. If it has downloaded this data object before, or another data
object from the same storage location, then it could keep a history of download time and network speed for each location to improve the estimates.

Internally, the ExecutionPlanner could use these metrics to come up with an estimate for how long it thinks it would take to download the data. It can then report this in the offer by setting the `starttime` property to include a suitable delay.

If we have three ExecutionPlanner platforms at different distances from the data storage location, the client would see three offers, one starting now, one starting in 10 minuites, and one starting in 2 hours time. The different start times reflecting the different estimates for how long it would take to transfer the data.

All else being equal, the client would choose the offer with the earliest start time.

This gets us at least part way towards something like data proximity, without having to define new terms like `data-proximity` and `execution-cost`.

**Comment by Zarquan** [comment]
While we are thinking about these - would we want to include maximum and minimum values `data-proximity` and `execution-cost` in the datamodel. 

This would allow the client to specify it wanted a minimum score for  `data-proximity` and a maximum value for `execution-cost`, and the server to respond with the corresponding maximum `data-proximity` and minimum `execution-cost` in its offer.

**Comment by Zarquan** [comment]
If we rename `ivoa-obscore-resource` described above to `ivoa-obscore-votable`, then we could use `ivoa-obscore-resource` for a data resource that accepts parameters from the ObsCore data model to describe a single data object.

```
resources:
  data:
  - type: https://..../data-resource-types/ivoa-obscore-resource
    spec:

      # Input parameters supplied by the client.
      # A list of ObsCore values describing the data object.
      # See https://ivoa.net/documents/ObsCore
      service-params:
        obs_collection: !!str "...."
        obs_id: !!str "...."
        obs_publisher_did: !!str "...."
        access_url: !!str "...."

      # Response values in the offer supplied by the ExecutionPlanner.
      # A calculation of the data proximity.
      data-proximity: !!float
      # A calculation of the execution cost.
      execution-cost: !!float
```

**Comment by jesusjuansalgado** [comment]
Dave,  https://ivoa.net/documents/Notes/IVOAExecutionPlannerExtensions  looks to be a broken link. Could you please redirect me to the updated location?

I like the obscore example. I think older protocols (SIAP, SSAP, etc) are not our main target as these protocols are not so useful for missions with massive data where data location is more relevant. Yes, the metadata you mentioned is the one that could be used to determine if this is the correct execution node (based on input data)

BTW, we could mathematically define data-proximity using throughput (that could be valid for local data and streamed data) although I think the typical needed categories could be more ambiguous but still effective. The importance for many science platforms is to know if you have the data colocated, it is available (e.g. through a stream) or if the data used invalidates the workflow for this node (e.g. the only data used for this workflow is from SKA and the user is trying to use LSST resources)

**Comment by Zarquan** [comment]
https://ivoa.net/documents/Notes/IVOAExecutionPlannerExtensions hasn't been written yet.
It is where we would explain all the details of what `data-proximity` and `execution-cost` means and how they are calculated from the available ObsCore metadata.

**Comment by Zarquan** [comment]
To define data proximity using throughput, we would have to download some of the data ?

We don't want to download the whole of a 1TiB dataset, because we haven't decided if we want to execute the task on this platform yet. So I was thinking of a new IVOA capability, which would provide some test data to initialise a download speed test.

Alongside each storage location we could have a URL that returned a 1MiB file of random data that clients like ExecutionPlanner could download to make an initial estimate of download speed and data proximity ?

http://service.../download-test/random-1MiB

**Comment by Zarquan** [comment]
> The importance for many science platforms is to know if you have the data colocated

This is something we are finding with the Gaia data mining platform. Data access bandwidth is critical. For many machine learning use cases, co-located means really really close, e.g. in the same rack or direct attached on the same server.

So for a complex machine learning task we might want to specify a minimum bandwidth that the platform has to meet.

_"can you execute [this ML task] with [this data] available with a minimum [bandwidth]"_

**Comment by jesusjuansalgado** [comment]
yes, I was thinking the same about downloading the test data but you need to put this data in the remote location so I am not sure if we can do it (e.g. in the location of origin of the URLs from a TAP server and this could be out of the scope of the Execution Planner).

Another option could be a fractional download... you do not need to download the full data set but try to download a small fraction and abort (although that could be a little bit annoying for the servers). Maybe this produces open pipes at the server side until the download is discarded. I would need to investigate more. An interesting thing to think about... Maybe there are ways to do estimations as it looks a general problem.

I mentioned throughput and not bandwidth to cover colocated cases (where the data is available at the executed node directly by disk access) but, yes, your reasoning on constraints per bandwidth is interesting.

In any case, we could do a coarse-grained estimation of the data locality if we see it is too difficult to do a "perfect" calculation

**Comment by Zarquan** [comment]
> yes, I was thinking the same about downloading the test data but you need to put this data in the remote location so I am not sure if we can do it (e.g. in the location of origin of the URLs from a TAP server and this could be out of the scope of the Execution Planner).

An extra column in the TAP response, `data_access_test`, which points to the base of a well known set of file names for the test data ?

```
data_access_test=http://store21.example.com/test-data/
```

Gives access to test data downloads:
* http://store21.example.com/test-data/random-1KiB
* http://store21.example.com/test-data/random-10KiB
* http://store21.example.com/test-data/random-1MiB
* http://store21.example.com/test-data/random-10MiB

We could implement something like this in the SKA prototype to prove it works and then try convincing everyone in the IVOA to do the same.

**Comment by jesusjuansalgado** [comment]
yes, that is a good idea. The only problem is that implies changes in TAP, this is why I said it is out of the scope of the execution planner but it is, certainly, a very good recommendation for a TAP update
